### PR TITLE
geo: implement ST_PointFromGeoHash({text,int4}) -> geometry

### LINKED ISSUE #48820: geo: implement ST_PointFromGeoHash({text,int4}) -> geometry
Implement `ST_PointFromGeoHash` on arguments {text,int4}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_PointFromGeoHash.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48441 for an example PR.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-06-30T05:29:32Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by himanshuchawla009**
fixes: #48820
This PR adds a function to calculate the convert 
geohash representation to point representation,
imitating the PostGIS funcitonality.

if no precision is specified ST_PointFromGeoHash returns a point based on full precision of the input GeoHash string.
If precision is specified ST_PointFromGeoHash will use that many characters from the GeoHash to create the point.

Release note (sql change): Implements ST_PointFromGeoHash
which converts geohash to point

### CODE REVIEW

File: pkg/geo/decode.go
Comment (otan) [inline pkg/geo/decode.go:24]: nit: capital letter for comments

File: pkg/sql/logictest/testdata/logic_test/geospatial
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/geospatial:4269]: can you put these near `parse_tests`?

File: pkg/sql/sem/builtins/geo_builtins.go
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:1182]: nit: spell "geohash" as "GeoHash".
Also rephrase as "Return a Geometry point from a GeoHash string." -- I think this sentence is clearer :)

File: pkg/geo/decode.go
Comment (otan) [inline pkg/geo/decode.go:18]: nit: rename to GeoHashToPoint
I think this should be "converts a GeoHash to a Geometry Point using a Lng/Lat representation of the GeoHash".

**Review by otan — COMMENTED** [review]
Thanks for contributing!

I have a couple of minor comments.

File: pkg/sql/sem/builtins/geo_builtins.go
Comment (himanshuchawla009) [reply pkg/sql/sem/builtins/geo_builtins.go:1182]: can i rephrase as "Return a Geometry point from a GeoHash string with supplied precision." as this function also takes precision.

File: pkg/geo/decode.go
Comment (himanshuchawla009) [reply pkg/geo/decode.go]: done

File: pkg/sql/logictest/testdata/logic_test/geospatial
Comment (himanshuchawla009) [reply pkg/sql/logictest/testdata/logic_test/geospatial]: done

File: pkg/sql/sem/builtins/geo_builtins.go
Comment (himanshuchawla009) [reply pkg/sql/sem/builtins/geo_builtins.go]: done, let me know if we need to rephrase it.

File: pkg/geo/decode.go
Comment (himanshuchawla009) [reply pkg/geo/decode.go]: done

File: pkg/sql/logictest/testdata/logic_test/geospatial
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/geospatial:842]: you can omit the semicolons here `;`

File: pkg/geo/decode.go
Comment (otan) [inline pkg/geo/decode.go:1]: do you mind renaming this to `geohash.go`?

**Review by otan — COMMENTED** [review]
Sorry couple of small things left!

Also looks like this will fail CI because a few files need to be regenerated -- please type `make buildshort` and commit the file changes there!

File: pkg/sql/logictest/testdata/logic_test/geospatial
Comment (himanshuchawla009) [reply pkg/sql/logictest/testdata/logic_test/geospatial]: done

File: pkg/geo/decode.go
Comment (himanshuchawla009) [reply pkg/geo/decode.go]: done

File: pkg/sql/logictest/testdata/logic_test/geospatial
Comment (himanshuchawla009) [reply pkg/sql/logictest/testdata/logic_test/geospatial]: i have used make buildshort command , hopefully this will make CI work this time.

**Comment by otan** [comment]
looks like it added the extra *.pb.go files, do you mind reverting these changes?
(it's because you're running go1.14 when we run go1.13, you'll need to ensure these are reverted)

**Comment by himanshuchawla009** [comment]
> looks like it added the extra *.pb.go files, do you mind reverting these changes?
> (it's because you're running go1.14 when we run go1.13, you'll need to ensure these are reverted)

But these files have same code as before only the indentation is changed. I have checked in code difference.

**Comment by otan** [comment]
CI is strict on the files being the same as a clean generation -- i can't merge this unless it's green.

If you're curious on the error message, click details on
image

**Comment by himanshuchawla009** [comment]
> clean regeneration


> Unfortunately, CI is strict on the files being the same as a clean regeneration -- i can't merge this unless it's green.
> 
> If you're curious on the error message, click details on
> image

i think i am not able to resolve linting issues. There are 3 lint issues caught be CI pipeline:-
One to those is mentioned below, so i have rename the function name
`pkg/geo/geohash.go:20:6: func name will be used as geo.GeoHashToPoint by other packages, and that stutters; consider calling this HashToPoint
`
For other 2 issues i am not very sure what i need to fix.

**Comment by otan** [comment]
the roachtest issue looks like a flake, you should be able to rebase and i think it should go away

**Comment by otan** [comment]
The `TestGofmtSimplify` looks like you'll need to run `crlfmt -w` on that file. Try
```
make bin/crlfmt
crlfmt -w -tab 2 <filename>
```

**Comment by himanshuchawla009** [comment]
> The `TestGofmtSimplify` looks like you'll need to run `crlfmt -w` on that file. Try
> 
> ```
> make bin/crlfmt
> crlfmt -w -tab 2 <filename>
> ```

no luck,

**Comment by otan** [comment]
In that case you can try run gofmt -s on that file, or apply the changes by
hand.

**Comment by otan** [comment]
(i can take care of this for you if you want! your name will still be on the commit :])

**Comment by himanshuchawla009** [comment]
> (i can take care of this for you if you want! your name will still be on the commit :])

yes, plz if possible... in upcoming commits i will take care of these things.

**Comment by otan** [comment]
Sorry looks like our CI is flaky atm. I'll try tomorrow.
This is a sub-par experience and apologise for the trouble.

**Comment by himanshuchawla009** [comment]
> Sorry looks like our CI is flaky atm. I'll try tomorrow.
> This is a sub-par experience and apologise for the trouble.

Thanks a lot for helping me out with this, looks like now CI has given green signal and all issues has been resolved.

**Review by otan — APPROVED** [review]

**Comment by Amruta-Ranade** [comment]
@himanshuchawla009 Thank you for contributing to CockroachDB this year. As a token of our appreciation, we would like to send you a gift. Please DM me on our [community Slack](https://cockroa.ch/slack) @Amruta so I can send you a link. (If you don’t want a gift, we also have a charitable contribution choice.) All orders need to be in by December 13, so please contact me as soon as possible :)

**Pull request merged**
